File: server/src/main/java/edu/cnm/deepdive/coffeeshop/service/RatingService.java
package edu.cnm.deepdive.coffeeshop.service;

import edu.cnm.deepdive.coffeeshop.model.dto.RatingDto;
import edu.cnm.deepdive.coffeeshop.model.dto.RatingRequestDto;
import edu.cnm.deepdive.coffeeshop.model.dto.RatingUpdateRequestDto;
import edu.cnm.deepdive.coffeeshop.model.entity.Profile;
import edu.cnm.deepdive.coffeeshop.model.entity.Rating;
import java.util.List;
import java.util.UUID;

public interface RatingService {

  List<RatingDto> listRatings(UUID visitId);

  void deleteRating(UUID visitId, UUID interest, Profile profile);

  RatingDto saveRating(UUID visitId, RatingRequestDto ratingRequestDto, Profile profile);

  RatingDto updateRating(UUID visitId, UUID interestId, RatingUpdateRequestDto ratingUpdateRequestDto, Profile profile);

  RatingDto getRating(UUID visitId, UUID interestId);

}

File: app/src/main/java/edu/cnm/deepdive/coffeeshop/service/RatingService.kt
package edu.cnm.deepdive.coffeeshop.service

import edu.cnm.deepdive.coffeeshop.model.domain.Interest
import edu.cnm.deepdive.coffeeshop.model.domain.Visit
import edu.cnm.deepdive.coffeeshop.model.dto.openapi.RatingDto
import edu.cnm.deepdive.coffeeshop.model.dto.openapi.RatingRequestDto
import edu.cnm.deepdive.coffeeshop.repository.RatingRepository
import jakarta.inject.Inject
import jakarta.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.future.future

@Singleton
class RatingService @Inject constructor(private val ratingRepository: RatingRepository){

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    suspend fun addRating(visit: Visit, ratingRequestDto: RatingRequestDto) =
        scope.future {
            ratingRepository.addRating(visit, ratingRequestDto)
        }

    suspend fun removeRating(visit: Visit, interest: Interest) =
        scope.future {
            ratingRepository.removeRating(visit, interest)
        }

    suspend fun getRating(visit: Visit) =
        scope.future {
            ratingRepository.getRatings(visit)
        }

}